File: rtls-kmp/src/commonMain/kotlin/com/rtls/kmp/SyncPullAPI.kt
package com.rtls.kmp

/**
 * Opaque token from server for incremental fetch. Client persists and passes on next fetch.
 */
data class SyncCursor(val value: ByteArray) {
    constructor(string: String) : this(string.encodeToByteArray())

    fun stringValue(): String = value.decodeToString()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SyncCursor) return false
        return value.contentEquals(other.value)
    }

    override fun hashCode(): Int = value.contentHashCode()
}

/**
 * Result of a pull (fetch) from server.
 */
data class SyncFetchResult(
    val items: List<LocationPoint>,
    val nextCursor: SyncCursor? = null,
    val serverTimeMs: Long? = null
)

/**
 * Optional: implement for bidirectional sync. If not provided, engine is upload-only.
 */
interface LocationPullAPI {
    suspend fun fetch(since: SyncCursor?): SyncFetchResult
}

File: Sources/RTLSCore/SyncPullAPI.swift
import Foundation

/// Opaque token from server for incremental fetch. Client persists and passes on next fetch.
public struct SyncCursor: Sendable, Equatable {
    public let value: Data

    public init(value: Data) {
        self.value = value
    }

    public init(string: String) {
        self.value = Data(string.utf8)
    }

    public var stringValue: String? {
        String(data: value, encoding: .utf8)
    }
}

/// Result of a pull (fetch) from server.
public struct SyncFetchResult: Sendable {
    public var items: [LocationPoint]
    public var nextCursor: SyncCursor?
    public var serverTime: Date?

    public init(items: [LocationPoint], nextCursor: SyncCursor? = nil, serverTime: Date? = nil) {
        self.items = items
        self.nextCursor = nextCursor
        self.serverTime = serverTime
    }
}

/// Optional: implement for bidirectional sync. If not provided, engine is upload-only.
public protocol LocationPullAPI: Sendable {
    /// Fetch server changes since last known cursor. Returns items and next cursor.
    func fetch(since cursor: SyncCursor?) async throws -> SyncFetchResult
}
